### PR TITLE
Add TMC2209 support

### PR DESCRIPTION
**Pull request by thinkyhead**
Supersedes #14215 by @bigtreetech

---
### Description
Add support for Trinamic TMC2209 stepper drivers.

### Benefits
  1. Add stallGuard detection function.
  2. Add slave address. TMC2209 slave address can be set from 0x00 to 0x03 (up to 4 TMC2209s per UART port).

### Considerations
This version needs the TMC driver library version >=0.3.6. See [TMC libraries update](https://github.com/teemuatlut/TMCStepper/pull/37)

### CODE REVIEW

**Comment by teemuatlut** [comment]
@thinkyhead Would you like me to review or adopt the PR?

**Comment by thinkyhead** [comment]
Once it's passing Travis CI, I would definitely appreciate the additional review!

File: Marlin/src/feature/tmc_util.cpp
Comment (thinkyhead) [inline Marlin/src/feature/tmc_util.cpp:990]: @teemuatlut — The `TMC2209Stepper` class doesn't have a `TCOOLTHRS` method. So I assume these two functions are incorrect.

Comment (teemuatlut) [reply Marlin/src/feature/tmc_util.cpp:990]: Actually it should have the method but I've just missed it. Adding it now but it'll be included in the next release.

Sorry this is taking a while for me to start reviewing but the current issues have taken priority.

Comment (thinkyhead) [reply Marlin/src/feature/tmc_util.cpp:990]: No problem. We've got plenty to keep us occupied in the meantime!

**Comment by teemuatlut** [comment]
@thinkyhead Please see [teemuatlut/tmc2209](https://github.com/teemuatlut/Marlin/tree/tmc2209) / https://github.com/MarlinFirmware/Marlin/compare/bugfix-2.0.x...teemuatlut:tmc2209

There was quite a lot that I would've done differently from the branch this one is based one but I tried to retain whatever I liked.
The main differences are less copied code (much can be cast into parent to reuse), added some items to `M122` with TMC2209 and a different method for checking against SG limits.
Waiting for TMCStepper-v0.4.2 is not necessary and it will likely not add anything to TMC2209 support.

**Comment by thinkyhead** [comment]
Thanks Teemu. I grabbed your branch and made it the new base for this PR, and then applied a few changes based on some elements of the previous PR. Please check this over to make sure I haven't messed it up too badly, and if it passes muster then I'll be happy to merge it later today.

**Comment by thinkyhead** [comment]
With TMCStepper v0.4.3 we get an error in this PR…
```
no matching function for call to 'TMCMarlin<TMC2209Stepper, 'X', '0', (AxisEnum)0u>::SGTHRS()'
inline int16_t homing_threshold() { return TMC2209Stepper::SGTHRS(); }
```
…because `SGTHRS` is a write method. I have added a commit that *maybe* fixes it…?

**Comment by teemuatlut** [comment]
I added it yesterday: https://github.com/teemuatlut/TMCStepper/commit/bd5cba6a0bf5f439746a295e5acd2a515570cddb

Another discussion here https://github.com/teemuatlut/Marlin/issues/30#issuecomment-502974700 would suggest the build will compile after that.

**Comment by thinkyhead** [comment]
Hmm… Travis must not be picking it up yet.

**Comment by teemuatlut** [comment]
You'd need to pull `master` as it's not part of a release yet.

`TMC2209` does not have `sgt` method.

**Comment by thinkyhead** [comment]
Ah, ok. We can wait until there's a release to merge this then.

**Comment by teemuatlut** [comment]
Or if you want you can point to master to see if it compiles and is ready to go otherwise. Then when the release is online (I can push the tag today) you can drop the commit before merging.

**Comment by thinkyhead** [comment]
Ok, I'm off to sleep for a few hours. This will surely be merged later today!

**Review by teemuatlut — CHANGES_REQUESTED** [review]
5c3cd93541d20df9326b9038b3b574476cbb4480 This commit can be dropped =)

**Comment by Hedda** [comment]
FYI; both the FYSETC Cheetah board and BIQU BIGTREETECH SKR MINI-E3 boards are now on Aliexpress

https://www.reddit.com/r/3Dprinting/comments/c1lttc/fysetc_cheetah_board_chinaclone_of_th3d_ezboard/

https://www.reddit.com/r/BIGTREETECH/comments/c0hdht/bigtreetech_skr_mini_e3_done/

Both these boards feature integrated TMC2209 drivers and STM32F103 32-bit based MCU

**Comment by teemuatlut** [comment]
https://github.com/teemuatlut/TMCStepper/releases/tag/v0.4.4

* Added TMC2209::SGTHRS()

**Comment by Gerggg** [comment]
Does the addition of the TMC2209 correct the compatibility problem we encounter between TMC2208 and linear advance? as a reminder, the extruder motor shuts off when TMC2208 is used with Marlin's Linear Advance feature.

**Comment by coltonmccormack** [comment]
@Gerggg That was already addressed with #14195 . Try enabling SQUARE_WAVE_STEPPING.

**Comment by Gerggg** [comment]
Thank you, I'll try this.

**Comment by Gerggg** [comment]
I see that Travis has validated, what is the next step before passing the PR?

**Comment by hhoweson** [comment]
Has this been merged yet?

**Comment by thinkyhead** [comment]
> …what is the next step…

We're waiting for TMCStepper 0.4.5 to propagate so that PlatformIO will automatically download it for the build.

**Comment by zztopper** [comment]
> > …what is the next step…
> 
> We're waiting for TMCStepper 0.4.5 to propagate so that PlatformIO will automatically download it for the build.

Travis is validating now, what's next?

**Comment by thinkyhead** [comment]
This is auspicious. 
image

**Comment by teemuatlut** [comment]
Just FYI to those following the thread here; I do not expect sensorless homing to work with TMC2209 at the moment.

**Comment by Grogyan** [comment]
@teemuatlut I know a lot of people are eager for this feature int TMC2209, hopefully this will be coming very soon

**Comment by drewzh** [comment]
Where can I track progress on the stallguard support on TMC2209? Many thanks for your hard efforts. I'm about to install these to replace my 2208's.

**Comment by teemuatlut** [comment]
Keep an eye on Marlin issue tracker and pull requests. My Marlin fork is likely to be the first to introduce any changes required but not necessarily if someone gets there first.
Then there's the TMCStepper library but I don't think that should require any changes with this one.
________________________________

**Comment by drewzh** [comment]
Great, thanks @teemuatlut. I'm actually experiencing a problem with the TMC2209's and the latest Marlin dev actually - they seem to be stepping half the amount they need to. So for e.g, I've had to double my Ender 3's steps values for X, Y, Z and extruder. Do you know why that may be? When I switched from the default board to TMC2208's, no changes were necessary. I'm not sure where to seek help regarding this, so I was hoping you could point out what I'm doing wrong!

**Comment by Gerggg** [comment]
I have the same problem, the drivers are in 16 steps but I have set as for 32 steps

**Comment by drewzh** [comment]
I've switched my micro steps to 1/8 (from 1/16) in the firmware and returned my step values back in the EEPROM. I'm not sure which is the best approach. I'm assuming this isn't an expected behaviour?

**Comment by coltonmccormack** [comment]
@drewzh @Gerggg Since this code is already merged you guys should [open up an issue](https://github.com/MarlinFirmware/Marlin/issues/new/choose) (preferably use the same one, so whoever creates one first) and seek support there. That way any fixes can reference the originating issue in the commit.

**Comment by thinkyhead** [comment]
> I'm assuming this isn't an expected behaviour

I see no changes in Marlin logic from the recent patches. It is likely to be something with TMCStepper.

**Comment by drewzh** [comment]
I hadn't seen your latest message @thinkyhead, I've gone ahead and raised this at https://github.com/MarlinFirmware/Marlin/issues/14361

I'll leave it be, on the off chance that it IS something in Marlin. I'll look to raising this at TMCStepper as well.

**Comment by OneOfEleven** [comment]
> 
> 
> Just FYI to those following the thread here; I do not expect sensorless homing to work with TMC2209 at the moment.

I notice you specify that the TMC capable drivers that can do sensorless homing do it in spreadCycle mode @teemuatlut, but from what I read the new TMC2209 does it in stealthChop mode (new stallGuard4).

**Comment by oechslein** [comment]
FYI: For me sensorless homing works with TMC2209 with the current Marlin 2.x!

**Comment by drewzh** [comment]
Any specific changes to the config you had to make @oechslein? Any caveats before anyone else tries it?

**Comment by oechslein** [comment]
[20190623_Marlin_2_SKR1.3_2209.zip](https://github.com/MarlinFirmware/Marlin/files/3318325/20190623_Marlin_2_SKR1.3_2209.zip)

I uploaded my config, I don't think that there is anything specific (maybe enabling  "SQUARE WAVE STEPPING"). You need to have latest Marlin.

**Comment by drewzh** [comment]
@oechslein I'm actually finding that homing is working backwards when using sensorless_homing (X to the right, Y to the front). My initial go to would be to enable X|Y_MIN_ENDSTOP_INVERTING, but there's some sanity checks in SanityCheck.h line 2036 that's stopping compilation. I can't seem to get the logic to work when playing with ENDSTOPPULLUPS.

The correct pin on the 2209 for the endstop is the pin to the left of the SEN text?

**Comment by oechslein** [comment]
@drewzh I'm using the SKRv1.3 there its just a jumper to attach xdiag to the 2209.
But if homing moves in the wrong direction than I would guess that simple movements (G1 X100) will also go in the wrong direction? Maybe try a `#define INVERT_X_DIR true`

**Comment by drewzh** [comment]
@oechslein I definitely have all those red jumpers removed (minus the ones to enable UART), so it sounds like I don't have it connected correctly. Normal movements react correctly actually, it's just homing that's not working as expected. Is it just XDIAG that needs shorting? I presume I still need to connect the diag1 pin of the 2209 to the end stops?

**Comment by drewzh** [comment]
@oechslein Thanks! I've connected all of the DIAG1 jumpers. After playing with sensitivity, I've landed on X100 and Y100 (completely different to what's required for TMC2130 - it seems higher values are more sensitive). Y axis homing is silent, however X axis hits the end stop with a good thud - so that's something I'll have to work out (unless that's normal - I've never done this before). Only other thing I've noticed is that my onboard speaker now beeps when turning the printer on. This was directly after connecting the DIAG1 jumpers.

**Comment by teemuatlut** [comment]
The 2209 stallGuard4 tuning works in the reverse direction (or rather the 2130 works in reverse. Depends on how you look at it).

2130:
* SG values are from -64..63
* Higher values mean less sensitive probing
* You can think of it like tuning on offset that's applied to the measurement, which is always compared to 0.

2209 (stallGuard4):
* SG values are from 0..255
* **Lower** values mean less sensitive
* The measurement is compared to a value that can be set by the user

**Comment by thisiskeithb** [comment]
> Lower values mean less sensitive

The `SENSORLESS_HOMING` block in Configuration_adv.h should probably be updated to reflect this.

**Comment by drewzh** [comment]
I'd be very interested in whether you're experiencing a harsher home on the X vs a soft home on the Y @oechslein. I've ruled out any hardware or config issues (sensitivity etc). For some reason my X axis just wants to hit and judder for a second.

**Comment by oechslein** [comment]
@drewzh 
Did you set X_HOME_BUMP_MM to zero?
And I would also suggest to enable HOMING_BACKOFF_MM and set it for x and y to 5 (or more).
My X sensitivity value is higher than my Y value, but both a quiet.
You could also try to decrease homing speed (and then re-adjust sensitivity).

**Comment by drewzh** [comment]
@oechslein Thanks for getting back to me. Unfortunately I've tried both those suggestions and neither improve it. It's weird how the Y stop is so quiet but the X stop is jarring. I also tried a bunch of changes discussed here: https://github.com/MarlinFirmware/Marlin/issues/12074 to no avail. I can run the axis all the way to the end stop with a solid stop at the end manually and the belt is plenty tight enough... I'm at a loss.

**Comment by drewzh** [comment]
Interestingly @oechslein I've figured out a way to silence the sensorless homing and that's to home each axis in sequence rather than at the same time. I have a feeling that something else is at play here and doing this is just masking the main issue, but I'm not sure what!

Anyway, here's the option I set to mask the issue for now... 
`#define CODEPENDENT_XY_HOMING`

**Comment by EvilGremlin** [comment]
Just got my batch of 2209s, will be happy to jump in testing! I have MK8 type direct extruder with 1.8 motors, mega2560 and mainly interested in getting real one-pin UART (modified SoftwareSerial) and linear advance working. Any starting points?

**Comment by TheNitek** [comment]
Anyone tried the Fysetc TMC2209 with sensorless homing yet? UART connection works well but the when trying sensorless homing on my SKR 1.3 they always show "triggered" no matter of the sensitivity

**Comment by comps** [comment]
From what I remember reading the datasheet, they use different value range for stallguard threshold, 0-255 instead of ie. -64 to +63 of TMC2130. I don't know if the tmc library compensated for that or not, but you might try that first.

The whole stallguard seems to be completely reworked (stopping the motor on stall automatically, not relying on DIAG or register), so I'm really surprised it works by default ... or surprised how quickly it was implemented in the library. :)

**Comment by davidcgu** [comment]
> Anyone tried the Fysetc TMC2209 with sensorless homing yet? UART connection works well but the when trying sensorless homing on my SKR 1.3 they always show "triggered" no matter of the sensitivity

What you did to make them work? I have an SKR v.13 and TMC2208 V2.0 from Fysetc and are even not communicating for me

**Comment by davidcgu** [comment]
How did you get them working on UART, did you had to make some soldering on the driver?¿

**Comment by Projectile007** [comment]
For UART on 2209 v2.1 Solder TX and RX pins together on topside of the Driver. For 2208 Fystec 2.0 solder the two pads underneath the driver closes to the center of the board. [https://forum.fysetc.com/d/115-tmc2209-v2-1-uart](url)

**Pull request merged**
